sharding/collation.go

```diff
@@ -5,43 +5,97 @@ import (
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto/sha3"
+	"github.com/ethereum/go-ethereum/rlp"
 )
 
 // Collation base struct.
 type Collation struct {
 	header       *CollationHeader
+	body         []byte
 	transactions []*types.Transaction
 }
 
 // CollationHeader base struct.
 type CollationHeader struct {
-	shardID           *big.Int        //the shard ID of the shard.
-	chunkRoot         *common.Hash    //the root of the chunk tree which identifies collation body.
-	period            *big.Int        //the period number in which collation to be included.
-	proposerAddress   *common.Address //address of the collation proposer.
-	proposerSignature []byte          //the proposer's signature for calculating collation hash.
+	data collationHeaderData
 }
 
-// Header returns the collation's header.
-func (c *Collation) Header() *CollationHeader { return c.header }
+type collationHeaderData struct {
+	ShardID           *big.Int        // the shard ID of the shard.
+	ChunkRoot         *common.Hash    // the root of the chunk tree which identifies collation body.
+	Period            *big.Int        // the period number in which collation to be Pncluded.
+	ProposerAddress   *common.Address // address of the collation proposer.
+	ProposerSignature []byte          // the proposer's signature for calculating collation hash.
+}
 
-// Transactions returns an array of tx's in the collation.
-func (c *Collation) Transactions() []*types.Transaction { return c.transactions }
+// NewCollationHeader initializes a collation header struct.
+func NewCollationHeader(shardID *big.Int, chunkRoot *common.Hash, period *big.Int, proposerAddress *common.Address, proposerSignature []byte) *CollationHeader {
+	data := collationHeaderData{
+		ShardID:           shardID,
+		ChunkRoot:         chunkRoot,
+		Period:            period,
+		ProposerAddress:   proposerAddress,
+		ProposerSignature: proposerSignature,
+	}
+	return &CollationHeader{data}
+}
+
+// Hash takes the keccak256 of the collation header's contents.
+func (h *CollationHeader) Hash() (hash common.Hash) {
+	hw := sha3.NewKeccak256()
+	rlp.Encode(hw, h)
+	hw.Sum(hash[:0])
+	return hash
+}
 
 // ShardID is the identifier for a shard.
```

Comment:
This comment is not exactly accurate. It should be something like **ShardID the collation belongs to**